include/dlaf/permutations/general/impl.h

```diff
@@ -0,0 +1,201 @@
+//     │  in_offset      in │    │     │         out │
+//     │ │                  │    │     └─►┌─┬─┬─┐    │
+//     │◄┘ ┌─┐   ┌─┐   ┌─┐  │    │        │c│a│b│    │
+//     │   │a│   │b│   │c│  │    │    ┌──►│ │ │ │    │
+//     │   │ │ ┌►│ │   │ │◄─┼────┼──┐ │   └─┴─┴─┘    │
+//     │   └─┘ │ └─┘   └─┘  │    │  │ │      ▲       │
+//     │       │            │    │sz.rows()  │       │
+//     │      sz.rows()     │    │           └─      │
+//     │                    │    │          sz.cols()│
+//     └────────────────────┘    └───────────────────┘
+//
+// Example row permutations with `perm_arr = [3, 1]`:
+//
+//             ┌─── in_offset
+//             │                   out_begin
+//     ┌───────▼────────────┐    ┌──┬────────────────┐
+//     │                 in │    │  │            out │
+//     │       ┌─────┐      │    │  └►┌─────┐        │
+//   1 │       │  a  │      │    │    │  b  │        │
+//     │       └─────┘      │    │    ├─────┤ ◄─┐    │
+//     │                    │    │    │  a  │   │    │
+//     │       ┌─────┐      │    │    └──▲──┘   │    │
+//   3 │       │  b  │      │    │       │      │    │
+//     │       └──▲──┘      │    │ sz.cols()    │    │
+//     │          │         │    │          sz.rows()│
+//     └──────────┼─────────┘    └───────────────────┘
+//                │
+//             sz.cols()
+//
+//
+// [1]: The portion of each input column or row is defined by the interval [in_offset, in_offset +
+//      sz.col()) or the interval [in_offset, in_offset + sz.row()) respectively.
+// [2]: The input submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `in_tiles`
+// [3]: The subregion is defined by `begin` and `sz`
+// [4]: The output submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `out_tiles`
+//
+// Note: `in_tiles` should be `const T` but to avoid extra allocations necessary for unwrapping
+//       `std::vector<shared_future<matrix::Tile<const T, D>>>` it is left as non-const
+template <class T, Coord coord>
+void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeType in_offset,
+                       const matrix::Distribution& distr, const SizeType* perm_arr,
+                       const std::vector<matrix::Tile<T, Device::CPU>>& in_tiles,
+                       const std::vector<matrix::Tile<T, Device::CPU>>& out_tiles) {
+  constexpr Coord ocrd = orthogonal(coord);
+  std::vector<SizeType> splits =
+      util::interleaveSplits(sz.get<ocrd>(), distr.blockSize().get<ocrd>(),
+                             distr.distanceToAdjacentTile<ocrd>(in_offset),
+                             distr.distanceToAdjacentTile<ocrd>(out_begin.get<ocrd>()));
+
+  // Parallelized over the number of permutated columns or rows
+  std::vector<SizeType> loop_arr(to_sizet(sz.get<coord>()));
+  std::iota(std::begin(loop_arr), std::end(loop_arr), 0);
+  pika::for_each(pika::execution::par, std::begin(loop_arr), std::end(loop_arr), [&](SizeType i_perm) {
```

Comment:
```suggestion
  pika::for_loop(pika::execution::par, to_sizet(0), to_sizet(sz.get<coord>()), [&](SizeType i_perm) {
```

Reply:
Alternatively a `counting_iterator` type-thing from boost or self-made would be more appropriate to use with `for_each` to avoid having to allocate a vector.